cmd/multiFaAcceleration/multiFaAcceleration.go

```diff
@@ -0,0 +1,504 @@
+	var searchSpace []bed.Bed
+	var referenceLength, i, j, threshold int
+	var bitArray []byte
+
+	//a couple of safety checks
+	if len(records) != 4 {
+		log.Fatalf("multiFaAcceleration accepts a multiFa file with 4 records, found %v.", len(records))
+	}
+	if len(records[1].Seq) != len(records[0].Seq) || len(records[2].Seq) != len(records[0].Seq) || len(records[3].Seq) != len(records[0].Seq) {
+		log.Fatalf("Error. All records must be of the same sequence length.")
+	}
+
+	referenceLength = fasta.AlnPosToRefPos(records[0], len(records[0].Seq)-1)
+
+	if s.SearchSpaceBed != "" {
+		//if we want to limit our search to a particular set of regions, we populate the bitArray with 1s instead of 0s at positions overlapping the input bed entries.
+		searchSpace = bed.Read(s.SearchSpaceBed)
+		bitArray = make([]byte, referenceLength)
+		for i = range searchSpace {
+			if searchSpace[i].Chrom == s.ChromName {
+				for j = searchSpace[i].ChromStart; j < searchSpace[i].ChromEnd; j++ {
+					bitArray[j] = 1
+				}
+			}
+		}
+		threshold = int(s.SearchSpaceProportion * float64(s.WindowSize)) //the minimum number of bases at which a window must overlap the search space in order to be considered a valid window.
+	}
+
+	var currDistances Distances
+	var distanceCache = make(map[Distances]BranchLengths)
+	var referenceCounter int = 0
+	var reachedEnd bool = false
+	var b1, b3 float64
+	var currCount int
+	var pass, containedInMap bool
+
+	//variables for normalization
+	var velSum, initialSum float64 = 0, 0
+	var branchCacheSlice = make([]BranchCache, 0)
+
+	for alignmentCounter := 0; reachedEnd == false && referenceCounter < referenceLength-s.WindowSize; alignmentCounter++ {
+		if s.Verbose && alignmentCounter%1000000 == 0 {
+			log.Printf("alignmentCounter: %v\n", alignmentCounter)
+		}
+		currCount, pass = thresholdCheckPasses(s, currCount, threshold, bitArray, referenceCounter) //first we see if we have a valid window.
+		if records[0].Seq[alignmentCounter] != dna.Gap {                                            //and if we are at a reference position.
+			if pass {
+				if s.UseSnpDistance {
+					reachedEnd = fourWaySnpDistances(records, alignmentCounter, s, &currDistances)
+				} else {
+					reachedEnd = fourWayMutationDistances(records, alignmentCounter, s, &currDistances)
+				}
+
+				if _, containedInMap = distanceCache[currDistances]; !containedInMap { //if this tree has not been seen before, calculate branch lengths
+					distanceCache[currDistances] = alternatingLeastSquares(currDistances, s)
+				}
+				//no our distances should be in the cache, and we can do a simple lookup.
+				b1 = distanceCache[currDistances].B1
+				b3 = distanceCache[currDistances].B3
+
+				if !reachedEnd { //if we haven't run out of chromosome, we have another valid window to appear in our output.
+					velSum += b1
+					initialSum += b3
+					branchCacheSlice = append(branchCacheSlice, BranchCache{referenceCounter, referenceCounter + s.WindowSize, b1, b3})
```

Comment:
I think I am not completely understanding the code - how does the sliding window work at the end of a chromosome when there are 4 species and their chromosomes are not the same length? When 1 species runs out of chromosome, does that mean no more windows will appear? And then because the multiFa format breaks the chromosome up into blocks, you repeat the sliding windows for each block?

Reply:
the when maf is converted to multi-fa format, each sequence is the same length, and insertions and deletions are noted by - in reference or one of the aligning sequences